### DISCUSSION TITLE
`Type.|` for creating Union types

### DISCUSSION THREAD

**Discussion by Bumppoman** [joeldrapper/literal#9] [Ideas]
It would be nice to be able to do

```ruby
class Foo
  extend Literal::Attributes

  attribute :bar, Symbol | String # creates Union(Symbol, String) behind the scenes
end
```

I realize this would require monkey-patching or refinements to be used with built-in types, but, is a neat feature in Crystal that I think would be a good fit here.

**Comment by joeldrapper** [comment]
Hey, thanks for the suggestion. I’d like to avoid monkey-patching `Object` with a `|` method, as this could lead to unexpected behaviour if anything else uses that method.

If you want to use a patch in your own project, I believe this should work:

```ruby
class Object
  def |(other)
    Literal::Types::UnionType.new(self, other)
  end
end

class Literal::Types::UnionType
  def |(other)
    @types << other
    self
  end
end
```

**Comment by Bumppoman** [comment]
I was thinking of a refinement that defines the method on Class along these lines:

```ruby
module UnionPipe
  refine Class do
    def |(other)
      Literal::Types::UnionType.new(self, other)
    end
  end
end
```

I’m going to play around with it though, thanks for the feedback!

**Reply by joeldrapper** [reply]
Refinements are quite slow and they only work in the scope of a file that's `using` them. Additionally, it’s not just classes but any object that’s a type in Literal.

For example, you could have a `_Union(:a, :b, :c)` which would be valid for only the three objects `:a`, `:b`, or `:c`.

Anything that responds to `===` with a truthy/falsy value is a type. So every object is a type. Procs are particularly interesting since they alias `===` to `call`.

**Reply by Bumppoman** [reply]
I would find it a reasonable limitation that the syntactic sugar only worked on class-level types or predefined `Literal::Types` types…with the fallback to manually create a Union type of other objects.

Interesting bit about refinements because the discussion seemed like they were improving the slowness issues in 3.2…perhaps those changes didn’t make it in or didn’t make much of a difference.  Not something I use generally so I haven’t had cause to try it yet!
